src/main/kotlin/no/nav/tilleggsstonader/sak/vedtak/boutgifter/BoutgifterBeregnYtelseSteg.kt

```diff
@@ -65,55 +74,58 @@ class BoutgifterBeregnYtelseSteg(
                 begrunnelse = vedtak.begrunnelse,
             ),
         )
-        tilkjentYtelseService.lagreTilkjentYtelse(
-            saksbehandling = saksbehandling,
-            andeler =
-                finnAndelTilkjentYtelse(
-                    saksbehandling,
-                    beregningsresultat,
-                ),
+        lagreTilkjentYtelse(saksbehandling, beregningsresultat)
+    }
+
+    private fun beregnOgLagreOpphør(
+        saksbehandling: Saksbehandling,
+        vedtak: OpphørBoutgifterRequest,
+    ) {
+        feilHvis(saksbehandling.forrigeIverksatteBehandlingId == null) {
+            "Opphør er et ugyldig vedtaksresultat fordi behandlingen er en førstegangsbehandling"
+        }
+        feilHvis(saksbehandling.revurderFra == null) {
+            "revurderFra-dato er påkrevd for opphør"
+        }
+        val forrigeVedtak = hentVedtak(saksbehandling.forrigeIverksatteBehandlingId)
+
+        opphørValideringService.validerVilkårperioder(saksbehandling)
+
+        opphørValideringService.validerVedtaksperioderAvkortetVedOpphør(
+            forrigeBehandlingsVedtaksperioder = forrigeVedtak.data.vedtaksperioder,
+            revurderFraDato = saksbehandling.revurderFra,
+        )
+
+        val avkortedeVedtaksperioder = avkortVedtaksperiodeVedOpphør(forrigeVedtak, saksbehandling.revurderFra)
+
+        val beregningsresultat = beregningService.beregn(saksbehandling, avkortedeVedtaksperioder, TypeVedtak.OPPHØR)
+
+        vedtakRepository.insert(
+            GeneriskVedtak(
+                behandlingId = saksbehandling.id,
+                type = TypeVedtak.OPPHØR,
+                data =
+                    OpphørBoutgifter(
+                        vedtaksperioder = avkortedeVedtaksperioder,
+                        beregningsresultat = beregningsresultat,
+                        årsaker = vedtak.årsakerOpphør,
+                        begrunnelse = vedtak.begrunnelse,
+                    ),
+                gitVersjon = Applikasjonsversjon.versjon,
+            ),
         )
```

Comment:
For innvilgelse er dette trukket ut til en hjelpemetode: `lagInnvilgetVedtak()`. Kunne vi gjort det samme her?

Reply:
Done!